### ISSUE TITLE
AssertionError on GetHomeDir()

### ISSUE THREAD

**Issue by MagicRB** [oilshell/oil#1073]
Oil shell crashes when I run it as PID 1 in initrd, I tried creating `/bin/sh` and `/usr/bin/env`, but nope. It crashes with https://gist.github.com/MagicRB/7f9b9c5a7afd2bb97fdcb8169c8f19c1. So it must be related to https://github.com/oilshell/oil/issues/47 but I can't figure out how to fix it.

**Comment by andychu** [comment]
Thanks for testing it and thanks for the report!


Do you have a repo that shows how you are starting Oil?   I'm not very familiar with `initrd`.




http://www.oilshell.org/blog/2017/05/05.html

**Comment by andychu** [comment]
(hit enter too soon)

BTW I think other people have tried this, and ran into problems, so I'm not that surprised you ran into problems.  But we should definitely try to make this work.

A few things

1. We're currently using the "Python app bundle", which should go away with `oil-native` (the C++ translation)

The app bundle isn't really documented, but it is mentioned here: http://www.oilshell.org/blog/2017/05/05.html

2. However this should work, and I'm not sure why it doesn't.

### Details on the App Bundle

There are two parts: native code (slice of CPython) and a `.zip` of bytecode in the **same file** `oil.ovm`.

The native code running in `oil.ovm` has to **find** itself as a data to open the `.zip` part!!!  This is hard in Unix.

It does this by opening `argv[0]`. 

OK so now I'm almost 100% sure I know what the bug is!!  `initrd` doesn't start the `oil.ovm` process like a shell would.

In the `exec()` system call, you can get `arg0` to whatever you want.  Let me see

**Comment by andychu** [comment]
So I think the difference comes here when we make a Python importer out of `argv[0]`

https://github.com/oilshell/oil/blob/master/Python-2.7.13/Modules/main.c#L253

So basically if that string isn't set to the "right" value so we can find the zip data in `oil.ovm`, we will get the error you observed.

----

This is actually a hard problem in Unix -- the reason that this doesn't happen in other shells is that none of them have to load byte code or data files

So this problem naturally goes away with `oil-native`

Nonetheless I think it would be nice to fix if we can.  But I'm not sure if there's a good way to do it ...

I welcome thoughts / investigation on this

**Comment by MagicRB** [comment]
That sounds plausible, but I now launch Oil from bash to setup `/bin/sh` and `/usr/bin/env`, I thought it was that.

```nix
let
pre-init = pkgs.writeShellScript "pre-init"
''
  export PATH=${pkgs.busybox}/bin
  # Borrowed from NixOS therefore it's licensed under the MIT license
  #### Activation script snippet usrbinenv:
  _localstatus=0
  mkdir -m 0755 -p /usr/bin
  ln -sfn ${pkgs.busybox}/bin/env /usr/bin/.env.tmp
  mv /usr/bin/.env.tmp /usr/bin/env # atomically replace /usr/bin/env
  # Create the required /bin/sh symlink; otherwise lots of things
  # (notably the system() function) won't work.
  mkdir -m 0755 -p /bin
  ln -sfn "${pkgs.busybox}/bin/sh" /bin/.sh.tmp
  mv /bin/.sh.tmp /bin/sh # atomically replace /bin/sh
  mkdir -pm 0777 /tmp
  mkdir -pm 0555 /var/empty

  ls / /bin /usr

  export PYTHONVERBOSE=8 OVM_VERBOSE=1

  exec ${main-init} 
'';

main-init = pkgs.writeOilScript
  { name = "init";
    text = builtins.readFile ./stages/1/init;
    substitutes =
    { inherit (pkgs) busybox;
    };
  };
in
  pre-init;
```
This is Nix, this bit should be readable to someone who's never used.

**Comment by andychu** [comment]
Ah OK so I think it comes down to the difference between this:

```
$ strace -e execve -ff bash -c 'exec ls'
execve("/bin/bash", ["bash", "-c", "exec ls"], 0x7ffe33df5698 /* 56 vars */) = 0
execve("/bin/ls", ["ls"], 0x56206905f8e0 /* 56 vars */) = 0
```

and this:

```
$ strace -e execve -ff bash -c 'exec -a foo ls'
execve("/bin/bash", ["bash", "-c", "exec -a foo ls"], 0x7ffc8e8ee238 /* 56 vars */) = 0
execve("/bin/ls", ["foo"], 0x5582e1f3b8e0 /* 56 vars */) = 0
```

So in the first case, we just do `exec ls`, and in the second case, we do `exec -a foo ls`, which is bash's way of controlling argv[0].

( `man execve` may be useful -- you can control the filename and argv[0] separately, but the process can only **see** argv[0]. )

So then:

- first case: the shell calls `/bin/ls`, and it sets argv[0] to `ls`.
- second case: the shell calls `/bin/ls`, and it sets argv[0] to `foo`.

So it would be nice to see what the text of `main-init` is ?  How is the shell being invoked?

It's possible that some kind of `-a` flag will fix it, although it's also possible that this is a hack, and something in Oil needs to be fixed.

However this is a fundamentally tricky issue because of the way Unix treats argv[0] and the fact that `oil.ovm` needs to find bytecode when starting.


(Seems related, but might not be: #743)

**Comment by andychu** [comment]
To be more specific, right now `argv[0]` needs to be set to a string that can be `open()`ed:

1. the absolute path to `oil.ovm`
2. a relative path to `oil.ovm`, which depends on the process's current working directroy (often `$PWD` in shell, although even that is nontrivial)

**Comment by MagicRB** [comment]
```nix
{ writeOilScript =                                                                             
    { name                                                                                     
    , text                                                                                     
    , substitutes ? {}                                                                         
    , path ? ""                                                                                
    , interpreter ? final.oil + "/bin/oil"                                                     
    }:                                                                                         
    final.writeScriptGen                                                                       
      { inherit name text substitutes path interpreter;                                        
      };                                                                                       
                                                                                               
                                                                                               
  writeOilScriptBin =                                                                          
    { name                                                                                     
    , text                                                                                     
    , substitutes ? {}                                                                         
    , path ? "/bin"                                                                            
    , interpreter ? final.oil + "/bin/oil"                                                     
    }:                                                                                         
    final.writeScriptGen                                                                       
      { inherit name text substitutes path interpreter;                                        
      };                                                                                       
                                                                                               
  writeScriptGen =                                                                             
    { name                                                                                     
    , text                                                                                     
    , substitutes ? {}                                                                         
    , path ? ""                                                                                
    , interpreter ? final.runtimeShell                                                         
    }:                                                                                         
    let outFile = if path == "" then "$out" else "$out/bin/${name}";                           
    in                                                                                         
      final.runCommandNoCC name                                                                
        ({                                                                                     
          nativeBuildInputs = [ final.busybox ];                                               
        } // substitutes)                                                                      
        ''                                                                                     
          TMPFILE=$(mktemp)                                                                    
          substituteAll ${final.writeText "${name}-orig" text} $TMPFILE                        
          ${if path == "" then "touch $out" else "mkdir -p $out/bin && touch $out/bin/${name}"}
          echo "#! ${interpreter}" > ${outFile}                                                
          cat $TMPFILE >> ${outFile}                                                           
          chmod +x ${outFile}                                                                  
        '';                                                                                    
};
```

**Comment by MagicRB** [comment]
thats the generator functions for the shell script, it basically just has a shebang at the top. As for seeing more, I'm testing Oil by attempting to a make a very paranoid extension to NixOS which would introduce a 3 or 2 linux bootloader boot process with TPM and Secure Boot support planned. Imagine the crypto nerds wet dream, well this is it.

**Comment by MagicRB** [comment]
I don't have oil in the path, maybe?

**Comment by MagicRB** [comment]
Also, I had Oil version 8.3 I think before and it actually worked, it broke after updating to 9.6

**Comment by MagicRB** [comment]
https://gist.github.com/MagicRB/fd4563678185137e6c2aef2f97bbcb74 also an strace, I'm at work rn, remotely but still, I'll dig through the strace in about an hour, you can do it in the meantime if you want

**Comment by MagicRB** [comment]
I'll try creating /etc/passed, it's the last file opened just before the crash

**Comment by MagicRB** [comment]
I guess this can now be closed, it was `/etc/passwd` and `/etc/nsswitch.conf`, the former has to have at least the root user but the latter can be completely empty. This script will do that:
```bash
mkdir -p /etc
touch /etc/passwd /etc/nsswitch.conf
echo "root:x:0:0:System administrator:/root:${pkgs.bashInteractive}/bin/bash" >> /etc/passwd
```

**Comment by andychu** [comment]
Hm does the missing file lead to `bin/osh` executing with the wrong argv[0]?  Or is it something else?

Or it looks like you are running bash with the root user?  That is required but can you now run osh with the non-root user?

Thanks for trying it and for the update

**Comment by MagicRB** [comment]
OK android 12 won't even allow me to edit `/etc/passwd` this is a major annoyance. Are you sure you're not checking permissions or something somewhere? Because something is accessing those files as can be seen in the strace, whatever it is its the reason for these issues.

**Comment by MagicRB** [comment]
OK android 12 won't even allow me to edit `/etc/passwd` this is a major annoyance. Are you sure you're not checking permissions or something somewhere? Because something is accessing those files as can be seen in the strace, whatever it is its the reason for these issues.

**Comment by andychu** [comment]
Hm I think the `/etc/passwd` thing is working around the issue but not the core issue.  I think the core issue is the argv[0] value ... 

If I could somehow run it in a VM it might be easier to track down.  This issue will naturally go away with `oil-native`, but I think other people may run into it in the meantime.

It can probably be reproduced outside of `initrd` pretty easily, although it will also help if we know how `initrd` starts the `osh` process, e.g. what the current working directory is and what argv[0] is.

**Comment by andychu** [comment]
Hm maybe there is some more going on, I am trying to trigger the bug outside of initrd with `exec -a`, and I can't do it ...

```
$ bash -c 'PATH=/bin exec -a osh /usr/local/bin/osh -c "echo hi"'
hi

$ bash -c 'PATH=/bin exec -a zzz /usr/local/bin/osh -c "echo hi"'
oil: Invalid applet name 'zzz'.
```

So I am starting osh when it's not in the $PATH, although maybe that's incidental too.

The core issue is that the native code running in osh has to `open()` itself in order to find the bytecode in the same file.

It's possible that it's better to put this in `~/.local/share/oil`  or something

**Comment by andychu** [comment]
Oops sorry I think I misdiagnosed this bug.  Well we both did, since it's not related to #47 

It probably is related to `/etc/passwd` since the AssertionError is


```
Traceback (most recent call last):
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 330, in _cpython_main_hook
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 303, in main
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 294, in AppBundleMain
  File "/home/andy/git/oilshell/oil/core/shell.py", line 559, in Main
AssertionError
FATAL: couldn't import from app bundle '/nix/store/m6g1cy3x9af6iyz45qxc9yckxgm2rb23-oil-0.9.6/bin/oil' (1)
Stripping the oil.ovm binary may cause this error.
See https://github.com/oilshell/oil/issues/47
```

Yeah so the error message is bad because it points to #47, sorry :-(

**Comment by andychu** [comment]
So I think what we need to do is:

- Fix this error message not to point to #47 , just say that it MAY be the cause
- See what other shells do when they can't get the home directory because `/etc/passwd` is messed up.  Rather than assert, there should be a fatal error!


Thanks for the report!

**Comment by MagicRB** [comment]
So it's `/etc/passed` missing, which causes oil not able to get the home Dir which then makes it fail?

**Comment by MagicRB** [comment]
Well bash works

**Comment by andychu** [comment]
This should have been fixed a couple months ago with #1578 

The latest release should have the fix - https://www.oilshell.org/release/0.17.0/
